### PR TITLE
add private helper functions for squaring Hamiltonian terms and colle…

### PR DESCRIPTION
**Pull request by anthayes92**
In order to evaluate flow constraints of the cost Hamiltonian in the max weighted cycle QAOA problem, we need to reliably expand quadratic terms and simplify the resulting expressions. 

Add a private function to find the coefficients and operators of the term of the Hamiltonian to be squared. 
Add a private function to collect duplicate terms of a Hamiltonian.

**Example of current and desired behaviour:**

Using the tensor product operation `@` as currently implemented we find:

```
>>> coeffs = [0, -1, -1, 1, 1]
>>> ops = [qml.Identity(wires=[0]), qml.PauliZ(wires=[0]), qml.PauliZ(wires=[1]), qml.PauliZ(wires=[2]), qml.PauliZ(wires=[4])]
>>> H = qml.Hamiltonian(coeffs, ops)
>>> square_H = H @ H
>>> print(square_H)
  (0) [I0]
+ (1) [Z0 Z0]
+ (1) [Z1 Z1]
+ (1) [Z2 Z2]
+ (1) [Z4 Z4]
+ (-2) [Z0 Z2]
+ (-2) [Z0 Z4]
+ (-2) [Z1 Z2]
+ (-2) [Z1 Z4]
+ (2) [Z0 Z1]
+ (2) [Z2 Z4]
```

The desired output of the above example would be: 
```
  (4) [I0]
+ (-2) [Z0 Z2]
+ (-2) [Z0 Z4]
+ (-2) [Z1 Z2]
+ (-2) [Z1 Z4]
+ (2) [Z0 Z1]
+ (2) [Z2 Z4]
```

This may no longer be needed when related issues on [Hamiltonian simplification](https://github.com/PennyLaneAI/pennylane/issues/1107) are closed

### CODE REVIEW

**Comment by trbromley** [comment]
Thanks @anthayes92! Looks good, could we come up with two minimal working examples that motivate why we need these functions (to be sure it's worth adding them in)?

E.g., could we have a simple Hamiltonian `H` where squaring with `H @ H` doesn't work as intended? Likewise a Hamiltonian `H` where calling the `simplify()` method doesn't work as intended?

This will also help us understand the shortfalls of the current approach in `qml.Hamiltonian`. Also, it can help us decide whether we could just make a quick change to the `qml.Hamiltonian.simplify()` method instead of having these improvements just available in the `cycle` module.

File: pennylane/qaoa/cycle.py [hunk 1]
```diff
@@ -14,8 +14,8 @@
 r"""
 Functionality for finding the maximum weighted cycle of directed graphs.
 """
-
```
Comment (trbromley) [inline pennylane/qaoa/cycle.py:17]: Reminder for changelog :smile:

Comment (anthayes92) [reply pennylane/qaoa/cycle.py:17]: updated WIP entry here: 4f616048847885538c35a39ab38f216d8490e3d8

File: pennylane/qaoa/cycle.py [hunk 2]
```diff
@@ -272,3 +272,37 @@ def loss_hamiltonian(graph: nx.Graph) -> qml.Hamiltonian:
...
+) -> Tuple[List[float], List[qml.operation.Observable]]:
+    """Calculates the coefficients and observables that compose the squared Hamiltonian.
+
+    Args:
+        coeffs (Iterable[float]): coeffients of the input Hamiltonian
+        ops (Iterable[qml.operation.Observable]): observables of the input Hamiltonian
+
+    Returns:
+        Tuple[List[float], List[qml.operation.Observable]]: The list of coefficients and list of observables
+            of the squared Hamiltonian.
```
Comment (trbromley) [inline pennylane/qaoa/cycle.py:288]:
```suggestion
        of the squared Hamiltonian.
```

**Review by trbromley — APPROVED** [review]
Awesome!

**Pull request merged**